src/Npgsql/Replication/ReplicationConnection.cs

```diff
@@ -428,74 +421,100 @@ async Task<TimelineHistoryFile> TimelineHistoryInternal()
+                        if (unchecked((ulong)Interlocked.Read(ref _lastReceivedLsn)) < endLsn)
+                            Interlocked.Exchange(ref _lastReceivedLsn, unchecked((long)endLsn));
+
+                        // dataLen = msg.Length - (code = 1 + walStart = 8 + walEnd = 8 + serverClock = 8)
+                        var dataLen = messageLength - 25;
+                        columnStream.Init(dataLen, canSeek: false);
+
+                        _cachedXLogDataMessage.Populate(new NpgsqlLogSequenceNumber(startLsn), new NpgsqlLogSequenceNumber(endLsn),
+                            sendTime, columnStream);
+                        yield return _cachedXLogDataMessage;
+
+                        // Our consumer may not have read the stream to the end, but it might as well have been us
+                        // ourselves bypassing the stream and reading directly from the buffer in StartReplication()
+                        if (!columnStream.IsDisposed && columnStream.Position < columnStream.Length && !bypassingStream)
+                            await buf.Skip(columnStream.Length - columnStream.Position, true, CancellationToken.None);
+
+                        continue;
+                    }
 
-                    continue;
-                }
+                    case 'k': // Primary keepalive message
+                    {
+                        await buf.Ensure(17, true, CancellationToken.None);
+                        var endLsn = buf.ReadUInt64();
+                        var timestamp = buf.ReadInt64();
+                        var replyRequested = buf.ReadByte() == 1;
+                        if (unchecked((ulong)Interlocked.Read(ref _lastReceivedLsn)) < endLsn)
+                            Interlocked.Exchange(ref _lastReceivedLsn, unchecked((long)endLsn));
 
-                case 'k': // Primary keepalive message
-                {
-                    await buf.Ensure(17, true, CancellationToken.None);
-                    var endLsn = buf.ReadUInt64();
-                    var timestamp = buf.ReadInt64();
-                    var replyRequested = buf.ReadByte() == 1;
-                    if (unchecked((ulong)Interlocked.Read(ref _lastReceivedLsn)) < endLsn)
-                        Interlocked.Exchange(ref _lastReceivedLsn, unchecked((long)endLsn));
+                        if (replyRequested)
+                            await SendFeedback(waitOnSemaphore: true, cancellationToken: CancellationToken.None);
 
-                    if (replyRequested)
-                        await SendFeedback(waitOnSemaphore: true, cancellationToken: CancellationToken.None);
+                        continue;
+                    }
 
-                    continue;
+                    default:
+                        throw connector.Break(new NpgsqlException($"Unknown replication message code '{code}'"));
+                    }
                 }
+            }
+            finally
```

Comment:
Just to clarify - if we've failed somewhere in a user code - does it still get called?

Reply:
Whatever happens in user code, this method will still need to exit somehow (i.e. with an exception). So if user code somehow causes the socket to break or something, we'd also end up here - right?

Reply:
@vonzshik am merging this to move on with other stuff, but let's continue this conversation if necessary - we can fix after this PR if needed.

Reply:
Just checked - seems to work fine.